libs/utils/include/celix_filter.h

```diff
@@ -44,63 +89,121 @@ typedef enum celix_filter_operand_enum
 
-CELIX_UTILS_EXPORT bool celix_filter_matchFilter(const celix_filter_t *filter1, const celix_filter_t *filter2);
+/**
+ * @brief Check whether the 2 filters are equal.
+ *
+ * Note that the equals is done based on the parsed filter string and not on the provided filter strings.
+ * So the filters parsed from the strings "(key=value)" and " (key=value)" are equal.
+ *
+ * @param[in] filter1 The first filter.
+ * @param[in] filter2 The second filter.
+ * @return True if the filters are equal, false otherwise.
+ */
+CELIX_UTILS_EXPORT bool celix_filter_equals(const celix_filter_t* filter1, const celix_filter_t* filter2);
 
-CELIX_UTILS_EXPORT const char* celix_filter_getFilterString(const celix_filter_t *filter);
+/**
+ * @brief Get the filter string representation.
+ *
+ * @param[in] filter The filter.
+ * @return The filter string representation. The returned string is owned by the filter, must not be freed and is only
+ *         valid as long as the filter is valid.
+ */
+CELIX_UTILS_EXPORT const char* celix_filter_getFilterString(const celix_filter_t* filter);
 
 /**
- * Find the filter attribute.
- * @return The attribute value or NULL if the attribute is not found.
+ * @brief Find the filter attribute.
+ *
+ * @param[in] filter The filter.
+ * @param[in] attribute The attribute to find.
+ * @return The found attribute value or NULL if the attribute is not found. The returned string is owned by the filter,
+ *         must not be freed and is only valid as long as the filter is valid.
  */
-CELIX_UTILS_EXPORT const char* celix_filter_findAttribute(const celix_filter_t *filter, const char *attribute);
+CELIX_UTILS_EXPORT const char* celix_filter_findAttribute(const celix_filter_t* filter, const char* attribute);
 
 /**
- * Check whether the filter indicates the mandatory presence of an attribute with a specific value for the provided attribute key.
+ * @brief Check whether the filter indicates the mandatory presence of an attribute with a specific value for the
+ * provided attribute key.
  *
  * Example:
  *   using this function for attribute key "key1" on filter "(key1=value1)" yields true.
  *   using this function for attribute key "key1" on filter "(!(key1=value1))" yields false.
  *   using this function for attribute key "key1" on filter "(key1>=value1)" yields false.
  *   using this function for attribute key "key1" on filter "(|(key1=value1)(key2=value2))" yields false.
-*/
-CELIX_UTILS_EXPORT bool celix_filter_hasMandatoryEqualsValueAttribute(const celix_filter_t *filter, const char *attribute);
+ *
+ * @param[in] filter The filter.
+ * @param[in] attribute The attribute to check.
+ * @return True if the filter indicates the mandatory presence of an attribute with a specific value for the provided
+ *         attribute key, false otherwise.
+ */
+CELIX_UTILS_EXPORT bool celix_filter_hasMandatoryEqualsValueAttribute(const celix_filter_t* filter,
```

Comment:
The function `celix_filter_hasMandatoryEqualsValueAttribute` (and its C++ variant) checks f a filter has a mandatory 'equals' attribute with the provided attribute name. This is hard to describe, but I updated the doxygen document. Hopefully this is now more clear. 

This can be used in the service-on-demand pattern to check if a requested filter uses certain attributes.
e.g. For a gauge_metric service check if the requesting filter provides a `name` and `unit_of_measurement` equals attribute.

Reply:
The comment says `using this function for attribute key "key1" on filter "(|(key1=value1)(key2=value2))" yields false`, which seems inconsistent with `EXPECT_TRUE(filter2.hasMandatoryEqualsValueAttribute("key3"))` for `filter2` I gave above. Note that `(!(&(key=value)(!(key3=value))))` is equivalent to `(|(!(key=value))(key3=value))`.